client/components/tinymce/plugins/mentions/suggestion.jsx

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+const Suggestion = React.createClass( {
+	displayName: 'Suggestion',
+
+	propTypes: {
+		avatarUrl: React.PropTypes.string,
+		username: React.PropTypes.string,
+		fullName: React.PropTypes.string,
+		query: React.PropTypes.string
+	},
+
+	getDefaultProps() {
+		return {
+			avatarUrl: '',
+			username: '',
+			fullName: '',
+			query: ''
+		};
+	},
+
+	getRegExpFor( type, textToHighlight ) {
+		const expressions = {};
+
+		expressions.username = '(^' + textToHighlight + ')(\\w*)\\s*';
+		expressions.fullName = '(^.*?)(\\b' + textToHighlight + ')(.*)';
+
+		return new RegExp( expressions[ type ], 'ig' );
+	},
+
+	highlight( content, textToHighlight, type ) {
+		const matcher = this.getRegExpFor( type, textToHighlight ),
+			matches = matcher.exec( content );
+
+		if ( matches ) {
+			const highlights = [];
+			let highlighted = false;
+
+			for ( let i = 1, length = matches.length; i < length; i++ ) {
+				let item = matches[ i ];
+
+				if ( textToHighlight.toLowerCase() === item.toLowerCase() && ! highlighted ) {
+					item = <strong key={ i }>{matches[ i ]}</strong>;
+					highlighted = true;
+				}
+
+				highlights.push( item );
```

Comment:
since the value of the next iteration depends on some carry from the previous iteration, this takes some extra steps. it can be done with `map` and `filter`, but [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) is uniquely designed for it as well. `reduce` iterates over a list and passes as input to the next iteration the output from the previous iteration, finally spitting out the result at the end. We provide the _seed_ or _initial value_ after giving a function for the iteration.

```js
const searchText = textToHighlight.toLowerCase()

if ( ! matches ) {
	return [ content ];
}

return matches
	.reduce( ( [ highlights, alreadyFound ], item, index ) => ( // destructure "carry"
		! alreadyFound && searchText === item.toLowerCase()
			? [ [ ...highlights, <strong key={ index }>{ item }</strong> ], true ]
			: [ [ ...highlights, item ], false ]
	), [ [], false ] ) // start with an empty list and "no we haven't found it yet"
	.shift(); // take the first item; we don't care about `alreadyFound` anymore
```

by the way, the function inside of `reduce` can be refactored too without giving up on the concept if it seems too unclear.